### PR TITLE
Add support for Gradient Clipping (clamp) in RNNT Numba loss

### PR DESCRIPTION
**Pull request by titu1994**
# Changelog
- Add support for `clamp` kwarg in rnnt loss numba. 
- Update configs to support this argument.

# Note
- This is different from using Pytorch Lightning `gradient_clip_val>0` as that will clip all gradients of the entire network by the limits of the clamp, this will only clip the gradient of the joint tensor itself and not the rest of the network.

### CODE REVIEW

File: nemo/collections/asr/parts/numba/rnnt_loss/rnnt_numpy.py [hunk 1]
```diff
@@ -90,6 +90,25 @@ def _assert_no_grad(tensor):
     )
 
 
+class LogSoftmaxGradModification(Function):
```
Comment (VahidooX) [inline nemo/collections/asr/parts/numba/rnnt_loss/rnnt_numpy.py:93]: Can this class be used for any activation? In yes, how about giving it a more general name as I do not see anything specific to logsoftmax?

File: nemo/collections/asr/parts/numba/rnnt_loss/utils/cpu_utils/cpu_rnnt.py [hunk 2]
```diff
@@ -137,6 +138,29 @@ def setup_probs(
                     self.log_probs2[offset + 1] = log_probs[idx(t, u, labels[u])]
 
 
+class LogSoftmaxGradModification(Function):
```
Comment (VahidooX) [inline nemo/collections/asr/parts/numba/rnnt_loss/utils/cpu_utils/cpu_rnnt.py:141]: Isn't this function a duplicate of the one in rnnt_numpy.py?

**Review by VahidooX — DISMISSED** [review]
LGTM! tests looks great!
Just some minor comments.

File: nemo/collections/asr/parts/numba/rnnt_loss/rnnt_numpy.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/parts/numba/rnnt_loss/rnnt_numpy.py:93]: Actually yeah you can use it to inject gradient update anywhere in the DAG but for now I want to keep this specific to LogSoftmax clamping because 1) it wastes memory - it copies the input joint tensor which is so expensive for rank 4 tensor 2) its forward backward is written specifically with clamping in mind. 

We can refactor this out (or remove it) if I can figure out how to do in place log softmax computation on CPU, but that will be a long term future task (if its even possible)..

File: nemo/collections/asr/parts/numba/rnnt_loss/utils/cpu_utils/cpu_rnnt.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/parts/numba/rnnt_loss/utils/cpu_utils/cpu_rnnt.py:141]: Yeah, for some weird reason im hitting circular dependency when I try to call the one from that class, so I duplicated here. I will be sending a new PR with enormous speed improvements for RNNT loss soon (maybe a few days) and I can refactor it out there.

**Review by titu1994 — COMMENTED** [review]
Thanks for comments !

File: examples/asr/conf/conformer/conformer_transducer_bpe.yaml [hunk 3]
```diff
@@ -182,6 +182,7 @@ model:
       # FastEmit regularization: https://arxiv.org/abs/2010.11148
       # You may enable FastEmit to reduce the latency of the model for streaming
       fastemit_lambda: 0.0  # Recommended values to be in range [1e-4, 1e-2], 0.001 is a good start.
+      clamp: -1.0  # if > 0, applies gradient clamping in range [-clamp, clamp].
```
Comment (krishnacpuvvada) [inline examples/asr/conf/conformer/conformer_transducer_bpe.yaml:185]: As mentioned in the PR notes, consider if it is worth hi-lighting the fact that "this will only clip the gradient of the joint tensor itself and not the rest of the network". (may be by further expanding the comment)

File: tests/collections/asr/numba/rnnt_loss/utils/test_gpu_rnnt_kernel.py [hunk 4]
```diff
@@ -167,3 +174,333 @@ def test_compute_betas_kernel(self):
...
+        )
+
+        # sync kernel
+        stream.synchronize()
+
+        # reshape grads
+        grads = grads.view([B, T, U, V])
+        diff = true_grads - grads[0].cpu().numpy()
+
+        assert np.abs(diff).mean() <= 1e-5
```
Comment (krishnacpuvvada) [inline tests/collections/asr/numba/rnnt_loss/utils/test_gpu_rnnt_kernel.py:505]: consider adding comment incase of false

File: tests/collections/asr/numba/rnnt_loss/test_rnnt_pytorch.py [hunk 5]
```diff
@@ -280,6 +280,84 @@ def test_case_large_random(self, device):
...
+        assert np.allclose(pt_grads, np_grads), "small_test gradient mismatch."
+
+    @pytest.mark.unit
+    @pytest.mark.parametrize('device', DEVICES)
+    @pytest.mark.parametrize('fastemit_lambda', [1.0, 0.01, 0.00001])
+    def test_case_small_fastemit_clamp(self, device, fastemit_lambda):
+        if device == 'cuda':
+            numba_utils.skip_numba_cuda_test_if_unsupported(__NUMBA_MINIMUM_VERSION__)
+
+        GRAD_CLAMP = 0.1
```
Comment (krishnacpuvvada) [inline tests/collections/asr/numba/rnnt_loss/test_rnnt_pytorch.py:337]: Purely for the sake of completeness, consider including a test case with -ve clamp val.

**Review by krishnacpuvvada — APPROVED** [review]
LGTM! (Just some minor comments)

File: tests/collections/asr/numba/rnnt_loss/test_rnnt_pytorch.py (hunk 5 above)
Comment (titu1994) [reply tests/collections/asr/numba/rnnt_loss/test_rnnt_pytorch.py:337]: Negatives are filtered out by abs(clamp) in the constructor, so it would change nothing at the kernel level. Good suggestion though, ill skip it for now.

File: examples/asr/conf/conformer/conformer_transducer_bpe.yaml (hunk 3 above)
Comment (titu1994) [reply examples/asr/conf/conformer/conformer_transducer_bpe.yaml:185]: Hmm, good point

Comment (titu1994) [reply examples/asr/conf/conformer/conformer_transducer_bpe.yaml]: Added a segment to state that it affects the joint only.

**Review by VahidooX — APPROVED** [review]
